### ISSUE TITLE
Progress bar does not reach end when i finish playing a video

### ISSUE THREAD

**Issue by qtiuto** [iina/iina#3331]
**System and IINA version:**

- macOS 11.2.1
- IINA 1.1.2 Build 128

When I play a mp4 to the end, the progress bar does not reach end with one frame gap. It 's obvious with pretty short video

**Comment by low-batt** [comment]
### Reproducing

As qtiuto stated the issue is obvious in short clips. I created the following file to reproduce the issue:
https://user-images.githubusercontent.com/86170219/141697007-df47cb57-c430-47f6-b073-63fa1448da0b.mp4

When this file is played not only does the thumb on the progress bar stop before it reaches the end as described by qtiuto, with the onscreen controller time labels configured to show milliseconds they too indicate play stopped short of the end of the file:
![issue-3331-iina-screen-shot](https://user-images.githubusercontent.com/86170219/141696991-e89d7401-35f9-4bf0-b5da-c2bc6f7f2630.jpg)

### Analysis

The values `videoDuration` and `videoPosition` in `PlaybackInfo` are used to determine the location of the thumb on the slider. The values of these two properties originate from the mpv properties `duration` and `time-pos`. The following debug logging was added to `MPVController.handlePropertyChange` to check the state of mpv properties when playback stops:

```swift
case MPVOption.PlaybackControl.pause:
  if let paused = UnsafePointer<Bool>(OpaquePointer(property.data))?.pointee {
    Logger.log("MPVOption.PlaybackControl.pause \(paused)")
    Logger.log("mpv property duration \(getDouble(MPVProperty.duration))")
    Logger.log("mpv property eof-reached \(getFlag(MPVProperty.eofReached))")
    Logger.log("mpv property estimated-frame-count \(getDouble(MPVProperty.estimatedFrameCount))")
    Logger.log("mpv property estimated-frame-number \(getDouble(MPVProperty.estimatedFrameNumber))")
    Logger.log("mpv property file-size \(getDouble(MPVProperty.fileSize))")
    Logger.log("mpv property percent-pos \(getDouble(MPVProperty.percentPos))")
    Logger.log("mpv property playback-time \(getDouble(MPVProperty.playbackTime))")
    Logger.log("mpv property playtime-remaining \(getDouble(MPVProperty.playtimeRemaining))")
    Logger.log("mpv property stream-end \(getDouble(MPVProperty.streamEnd))")
    Logger.log("mpv property stream-pos \(getDouble(MPVProperty.streamPos))")
    Logger.log("mpv property time-pos \(getDouble(MPVProperty.timePos))")
    Logger.log("mpv property time-remaining \(getDouble(MPVProperty.timeRemaining))")
```

This is what the logging showed when run using mpv 0.34.0:
```text
12:21:41.607 [iina][d] MPVOption.PlaybackControl.pause true
12:21:41.607 [iina][d] mpv property duration 1.001
12:21:41.607 [iina][d] mpv property eof-reached true
12:21:41.608 [iina][d] mpv property estimated-frame-count 29.0
12:21:41.608 [iina][d] mpv property estimated-frame-number 28.0
12:21:41.608 [iina][d] mpv property file-size 44613.0
12:21:41.608 [iina][d] mpv property percent-pos 96.66666666666669
12:21:41.608 [iina][d] mpv property playback-time 0.9676333333333333
12:21:41.608 [iina][d] mpv property playtime-remaining 0.033366666666666545
12:21:41.608 [iina][d] mpv property stream-end 44613.0
12:21:41.608 [iina][d] mpv property stream-pos 42149.0
12:21:41.608 [iina][d] mpv property time-pos 0.9676333333333333
12:21:41.608 [iina][d] mpv property time-remaining 0.033366666666666545
```

The property `eof-reached` is `true` and yet `percent-pos` is not 100%. The same values were seen when run against mpv 0.32.0.

Checking the frames of the video using ffmpeg:
<details>
<summary>Full ffprobe output</summary>

```text
low-batt@gag Movies$ ffprobe -count_frames -hide_banner -pretty -show_format -show_frames -show_streams issue-3331.mp4
Input #0, mov,mp4,m4a,3gp,3g2,mj2, from 'issue-3331.mp4':
  Metadata:
    major_brand     : isom
    minor_version   : 512
    compatible_brands: isomiso2avc1mp41
    encoder         : Lavf58.76.100
  Duration: 00:00:01.00, start: 0.000000, bitrate: 356 kb/s
  Stream #0:0(und): Video: h264 (High) (avc1 / 0x31637661), yuvj420p(pc, bt709), 640x360 [SAR 1:1 DAR 16:9], 346 kb/s, 29.97 fps, 29.97 tbr, 30k tbn, 59.94 tbc (default)
    Metadata:
      handler_name    : VideoHandler
      vendor_id       : [0][0][0][0]
[FRAME]
media_type=video
stream_index=0
key_frame=1
pkt_pts=0
pkt_pts_time=0:00:00.000000
pkt_dts=0
pkt_dts_time=0:00:00.000000
best_effort_timestamp=0
best_effort_timestamp_time=0:00:00.000000
pkt_duration=1001
pkt_duration_time=0:00:00.033367
pkt_pos=48
pkt_size=7.536133 Kibyte
width=640
height=360
pix_fmt=yuvj420p
sample_aspect_ratio=1:1
pict_type=I
coded_picture_number=0
display_picture_number=0
interlaced_frame=0
top_field_first=0
repeat_pict=0
color_range=pc
color_space=bt709
color_primaries=bt709
color_transfer=bt709
chroma_location=left
[SIDE_DATA]
side_data_type=H.26[45] User Data Unregistered SEI message
[/SIDE_DATA]
[/FRAME]
[FRAME]
media_type=video
stream_index=0
key_frame=0
pkt_pts=1001
pkt_pts_time=0:00:00.033367
pkt_dts=1001
pkt_dts_time=0:00:00.033367
best_effort_timestamp=1001
best_effort_timestamp_time=0:00:00.033367
pkt_duration=1001
pkt_duration_time=0:00:00.033367
pkt_pos=11354
pkt_size=837 byte
width=640
height=360
pix_fmt=yuvj420p
sample_aspect_ratio=1:1
pict_type=B
coded_picture_number=3
display_picture_number=0
interlaced_frame=0
top_field_first=0
repeat_pict=0
color_range=pc
color_space=bt709
color_primaries=bt709
color_transfer=bt709
chroma_location=left
[/FRAME]
[FRAME]
media_type=video
stream_index=0
key_frame=0
pkt_pts=2002
pkt_pts_time=0:00:00.066733
pkt_dts=2002
pkt_dts_time=0:00:00.066733
best_effort_timestamp=2002
best_effort_timestamp_time=0:00:00.066733
pkt_duration=1001
pkt_duration_time=0:00:00.033367
pkt_pos=10062
pkt_size=1.261719 Kibyte
width=640
height=360
pix_fmt=yuvj420p
sample_aspect_ratio=1:1
pict_type=B
coded_picture_number=2
display_picture_number=0
interlaced_frame=0
top_field_first=0
repeat_pict=0
color_range=pc
color_space=bt709
color_primaries=bt709
color_transfer=bt709
chroma_location=left
[/FRAME]
[FRAME]
media_type=video
stream_index=0
key_frame=0
pkt_pts=3003
pkt_pts_time=0:00:00.100100
pkt_dts=3003
pkt_dts_time=0:00:00.100100
best_effort_timestamp=3003
best_effort_timestamp_time=0:00:00.100100
pkt_duration=1001
pkt_duration_time=0:00:00.033367
pkt_pos=12191
pkt_size=950 byte
width=640
height=360
pix_fmt=yuvj420p
sample_aspect_ratio=1:1
pict_type=B
coded_picture_number=4
display_picture_number=0
interlaced_frame=0
top_field_first=0
repeat_pict=0
color_range=pc
color_space=bt709
color_primaries=bt709
color_transfer=bt709
chroma_location=left
[/FRAME]
[FRAME]
media_type=video
stream_index=0
key_frame=0
pkt_pts=4004
pkt_pts_time=0:00:00.133467
pkt_dts=4004
pkt_dts_time=0:00:00.133467
best_effort_timestamp=4004
best_effort_timestamp_time=0:00:00.133467
pkt_duration=1001
pkt_duration_time=0:00:00.033367
pkt_pos=7765
pkt_size=2.243164 Kibyte
width=640
height=360
pix_fmt=yuvj420p
sample_aspect_ratio=1:1
pict_type=P
coded_picture_number=1
display_picture_number=0
interlaced_frame=0
top_field_first=0
repeat_pict=0
color_range=pc
color_space=bt709
color_primaries=bt709
color_transfer=bt709
chroma_location=left
[/FRAME]
[FRAME]
media_type=video
stream_index=0
key_frame=0
pkt_pts=5005
pkt_pts_time=0:00:00.166833
pkt_dts=5005
pkt_dts_time=0:00:00.166833
best_effort_timestamp=5005
best_effort_timestamp_time=0:00:00.166833
pkt_duration=1001
pkt_duration_time=0:00:00.033367
pkt_pos=16349
pkt_size=1.087891 Kibyte
width=640
height=360
pix_fmt=yuvj420p
sample_aspect_ratio=1:1
pict_type=B
coded_picture_number=7
display_picture_number=0
interlaced_frame=0
top_field_first=0
repeat_pict=0
color_range=pc
color_space=bt709
color_primaries=bt709
color_transfer=bt709
chroma_location=left
[/FRAME]
[FRAME]
media_type=video
stream_index=0
key_frame=0
pkt_pts=6006
pkt_pts_time=0:00:00.200200
pkt_dts=6006
pkt_dts_time=0:00:00.200200
best_effort_timestamp=6006
best_effort_timestamp_time=0:00:00.200200
pkt_duration=1001
pkt_duration_time=0:00:00.033367
pkt_pos=15106
pkt_size=1.213867 Kibyte
width=640
height=360
pix_fmt=yuvj420p
sample_aspect_ratio=1:1
pict_type=B
coded_picture_number=6
display_picture_number=0
interlaced_frame=0
top_field_first=0
repeat_pict=0
color_range=pc
color_space=bt709
color_primaries=bt709
color_transfer=bt709
chroma_location=left
[/FRAME]
[FRAME]
media_type=video
stream_index=0
key_frame=0
pkt_pts=7007
pkt_pts_time=0:00:00.233567
pkt_dts=7007
pkt_dts_time=0:00:00.233567
best_effort_timestamp=7007
best_effort_timestamp_time=0:00:00.233567
pkt_duration=1001
pkt_duration_time=0:00:00.033367
pkt_pos=17463
pkt_size=957 byte
width=640
height=360
pix_fmt=yuvj420p
sample_aspect_ratio=1:1
pict_type=B
coded_picture_number=8
display_picture_number=0
interlaced_frame=0
top_field_first=0
repeat_pict=0
color_range=pc
color_space=bt709
color_primaries=bt709
color_transfer=bt709
chroma_location=left
[/FRAME]
[FRAME]
media_type=video
stream_index=0
key_frame=0
pkt_pts=8008
pkt_pts_time=0:00:00.266933
pkt_dts=8008
pkt_dts_time=0:00:00.266933
best_effort_timestamp=8008
best_effort_timestamp_time=0:00:00.266933
pkt_duration=1001
pkt_duration_time=0:00:00.033367
pkt_pos=13141
pkt_size=1.918945 Kibyte
width=640
height=360
pix_fmt=yuvj420p
sample_aspect_ratio=1:1
pict_type=P
coded_picture_number=5
display_picture_number=0
interlaced_frame=0
top_field_first=0
repeat_pict=0
color_range=pc
color_space=bt709
color_primaries=bt709
color_transfer=bt709
chroma_location=left
[/FRAME]
[FRAME]
media_type=video
stream_index=0
key_frame=0
pkt_pts=9009
pkt_pts_time=0:00:00.300300
pkt_dts=9009
pkt_dts_time=0:00:00.300300
best_effort_timestamp=9009
best_effort_timestamp_time=0:00:00.300300
pkt_duration=1001
pkt_duration_time=0:00:00.033367
pkt_pos=21426
pkt_size=839 byte
width=640
height=360
pix_fmt=yuvj420p
sample_aspect_ratio=1:1
pict_type=B
coded_picture_number=11
display_picture_number=0
interlaced_frame=0
top_field_first=0
repeat_pict=0
color_range=pc
color_space=bt709
color_primaries=bt709
color_transfer=bt709
chroma_location=left
[/FRAME]
[FRAME]
media_type=video
stream_index=0
key_frame=0
pkt_pts=10010
pkt_pts_time=0:00:00.333667
pkt_dts=10010
pkt_dts_time=0:00:00.333667
best_effort_timestamp=10010
best_effort_timestamp_time=0:00:00.333667
pkt_duration=1001
pkt_duration_time=0:00:00.033367
pkt_pos=20211
pkt_size=1.186523 Kibyte
width=640
height=360
pix_fmt=yuvj420p
sample_aspect_ratio=1:1
pict_type=B
coded_picture_number=10
display_picture_number=0
interlaced_frame=0
top_field_first=0
repeat_pict=0
color_range=pc
color_space=bt709
color_primaries=bt709
color_transfer=bt709
chroma_location=left
[/FRAME]
[FRAME]
media_type=video
stream_index=0
key_frame=0
pkt_pts=11011
pkt_pts_time=0:00:00.367033
pkt_dts=11011
pkt_dts_time=0:00:00.367033
best_effort_timestamp=11011
best_effort_timestamp_time=0:00:00.367033
pkt_duration=1001
pkt_duration_time=0:00:00.033367
pkt_pos=22265
pkt_size=830 byte
width=640
height=360
pix_fmt=yuvj420p
sample_aspect_ratio=1:1
pict_type=B
coded_picture_number=12
display_picture_number=0
interlaced_frame=0
top_field_first=0
repeat_pict=0
color_range=pc
color_space=bt709
color_primaries=bt709
color_transfer=bt709
chroma_location=left
[/FRAME]
[FRAME]
media_type=video
stream_index=0
key_frame=0
pkt_pts=12012
pkt_pts_time=0:00:00.400400
pkt_dts=12012
pkt_dts_time=0:00:00.400400
best_effort_timestamp=12012
best_effort_timestamp_time=0:00:00.400400
pkt_duration=1001
pkt_duration_time=0:00:00.033367
pkt_pos=18420
pkt_size=1.749023 Kibyte
width=640
height=360
pix_fmt=yuvj420p
sample_aspect_ratio=1:1
pict_type=P
coded_picture_number=9
display_picture_number=0
interlaced_frame=0
top_field_first=0
repeat_pict=0
color_range=pc
color_space=bt709
color_primaries=bt709
color_transfer=bt709
chroma_location=left
[/FRAME]
[FRAME]
media_type=video
stream_index=0
key_frame=0
pkt_pts=13013
pkt_pts_time=0:00:00.433767
pkt_dts=13013
pkt_dts_time=0:00:00.433767
best_effort_timestamp=13013
best_effort_timestamp_time=0:00:00.433767
pkt_duration=1001
pkt_duration_time=0:00:00.033367
pkt_pos=26643
pkt_size=930 byte
width=640
height=360
pix_fmt=yuvj420p
sample_aspect_ratio=1:1
pict_type=B
coded_picture_number=15
display_picture_number=0
interlaced_frame=0
top_field_first=0
repeat_pict=0
color_range=pc
color_space=bt709
color_primaries=bt709
color_transfer=bt709
chroma_location=left
[/FRAME]
[FRAME]
media_type=video
stream_index=0
key_frame=0
pkt_pts=14014
pkt_pts_time=0:00:00.467133
pkt_dts=14014
pkt_dts_time=0:00:00.467133
best_effort_timestamp=14014
best_effort_timestamp_time=0:00:00.467133
pkt_duration=1001
pkt_duration_time=0:00:00.033367
pkt_pos=25384
pkt_size=1.229492 Kibyte
width=640
height=360
pix_fmt=yuvj420p
sample_aspect_ratio=1:1
pict_type=B
coded_picture_number=14
display_picture_number=0
interlaced_frame=0
top_field_first=0
repeat_pict=0
color_range=pc
color_space=bt709
color_primaries=bt709
color_transfer=bt709
chroma_location=left
[/FRAME]
[FRAME]
media_type=video
stream_index=0
key_frame=0
pkt_pts=15015
pkt_pts_time=0:00:00.500500
pkt_dts=15015
pkt_dts_time=0:00:00.500500
best_effort_timestamp=15015
best_effort_timestamp_time=0:00:00.500500
pkt_duration=1001
pkt_duration_time=0:00:00.033367
pkt_pos=27573
pkt_size=839 byte
width=640
height=360
pix_fmt=yuvj420p
sample_aspect_ratio=1:1
pict_type=B
coded_picture_number=16
display_picture_number=0
interlaced_frame=0
top_field_first=0
repeat_pict=0
color_range=pc
color_space=bt709
color_primaries=bt709
color_transfer=bt709
chroma_location=left
[/FRAME]
[FRAME]
media_type=video
stream_index=0
key_frame=0
pkt_pts=16016
pkt_pts_time=0:00:00.533867
pkt_dts=16016
pkt_dts_time=0:00:00.533867
best_effort_timestamp=16016
best_effort_timestamp_time=0:00:00.533867
pkt_duration=1001
pkt_duration_time=0:00:00.033367
pkt_pos=23095
pkt_size=2.235352 Kibyte
width=640
height=360
pix_fmt=yuvj420p
sample_aspect_ratio=1:1
pict_type=P
coded_picture_number=13
display_picture_number=0
interlaced_frame=0
top_field_first=0
repeat_pict=0
color_range=pc
color_space=bt709
color_primaries=bt709
color_transfer=bt709
chroma_location=left
[/FRAME]
[FRAME]
media_type=video
stream_index=0
key_frame=0
pkt_pts=17017
pkt_pts_time=0:00:00.567233
pkt_dts=17017
pkt_dts_time=0:00:00.567233
best_effort_timestamp=17017
best_effort_timestamp_time=0:00:00.567233
pkt_duration=1001
pkt_duration_time=0:00:00.033367
pkt_pos=31360
pkt_size=928 byte
width=640
height=360
pix_fmt=yuvj420p
sample_aspect_ratio=1:1
pict_type=B
coded_picture_number=19
display_picture_number=0
interlaced_frame=0
top_field_first=0
repeat_pict=0
color_range=pc
color_space=bt709
color_primaries=bt709
color_transfer=bt709
chroma_location=left
[/FRAME]
[FRAME]
media_type=video
stream_index=0
key_frame=0
pkt_pts=18018
pkt_pts_time=0:00:00.600600
pkt_dts=18018
pkt_dts_time=0:00:00.600600
best_effort_timestamp=18018
best_effort_timestamp_time=0:00:00.600600
pkt_duration=1001
pkt_duration_time=0:00:00.033367
pkt_pos=30186
pkt_size=1.146484 Kibyte
width=640
height=360
pix_fmt=yuvj420p
sample_aspect_ratio=1:1
pict_type=B
coded_picture_number=18
display_picture_number=0
interlaced_frame=0
top_field_first=0
repeat_pict=0
color_range=pc
color_space=bt709
color_primaries=bt709
color_transfer=bt709
chroma_location=left
[/FRAME]
[FRAME]
media_type=video
stream_index=0
key_frame=0
pkt_pts=19019
pkt_pts_time=0:00:00.633967
pkt_dts=19019
pkt_dts_time=0:00:00.633967
best_effort_timestamp=19019
best_effort_timestamp_time=0:00:00.633967
pkt_duration=1001
pkt_duration_time=0:00:00.033367
pkt_pos=32288
pkt_size=895 byte
width=640
height=360
pix_fmt=yuvj420p
sample_aspect_ratio=1:1
pict_type=B
coded_picture_number=20
display_picture_number=0
interlaced_frame=0
top_field_first=0
repeat_pict=0
color_range=pc
color_space=bt709
color_primaries=bt709
color_transfer=bt709
chroma_location=left
[/FRAME]
[FRAME]
media_type=video
stream_index=0
key_frame=0
pkt_pts=20020
pkt_pts_time=0:00:00.667333
pkt_dts=20020
pkt_dts_time=0:00:00.667333
best_effort_timestamp=20020
best_effort_timestamp_time=0:00:00.667333
pkt_duration=1001
pkt_duration_time=0:00:00.033367
pkt_pos=28412
pkt_size=1.732422 Kibyte
width=640
height=360
pix_fmt=yuvj420p
sample_aspect_ratio=1:1
pict_type=P
coded_picture_number=17
display_picture_number=0
interlaced_frame=0
top_field_first=0
repeat_pict=0
color_range=pc
color_space=bt709
color_primaries=bt709
color_transfer=bt709
chroma_location=left
[/FRAME]
[FRAME]
media_type=video
stream_index=0
key_frame=0
pkt_pts=21021
pkt_pts_time=0:00:00.700700
pkt_dts=21021
pkt_dts_time=0:00:00.700700
best_effort_timestamp=21021
best_effort_timestamp_time=0:00:00.700700
pkt_duration=1001
pkt_duration_time=0:00:00.033367
pkt_pos=35767
pkt_size=983 byte
width=640
height=360
pix_fmt=yuvj420p
sample_aspect_ratio=1:1
pict_type=B
coded_picture_number=23
display_picture_number=0
interlaced_frame=0
top_field_first=0
repeat_pict=0
color_range=pc
color_space=bt709
color_primaries=bt709
color_transfer=bt709
chroma_location=left
[/FRAME]
[FRAME]
media_type=video
stream_index=0
key_frame=0
pkt_pts=22022
pkt_pts_time=0:00:00.734067
pkt_dts=22022
pkt_dts_time=0:00:00.734067
best_effort_timestamp=22022
best_effort_timestamp_time=0:00:00.734067
pkt_duration=1001
pkt_duration_time=0:00:00.033367
pkt_pos=34625
pkt_size=1.115234 Kibyte
width=640
height=360
pix_fmt=yuvj420p
sample_aspect_ratio=1:1
pict_type=B
coded_picture_number=22
display_picture_number=0
interlaced_frame=0
top_field_first=0
repeat_pict=0
color_range=pc
color_space=bt709
color_primaries=bt709
color_transfer=bt709
chroma_location=left
[/FRAME]
[FRAME]
media_type=video
stream_index=0
key_frame=0
pkt_pts=23023
pkt_pts_time=0:00:00.767433
pkt_dts=23023
pkt_dts_time=0:00:00.767433
best_effort_timestamp=23023
best_effort_timestamp_time=0:00:00.767433
pkt_duration=1001
pkt_duration_time=0:00:00.033367
pkt_pos=36750
pkt_size=743 byte
width=640
height=360
pix_fmt=yuvj420p
sample_aspect_ratio=1:1
pict_type=B
coded_picture_number=24
display_picture_number=0
interlaced_frame=0
top_field_first=0
repeat_pict=0
color_range=pc
color_space=bt709
color_primaries=bt709
color_transfer=bt709
chroma_location=left
[/FRAME]
[FRAME]
media_type=video
stream_index=0
key_frame=0
pkt_pts=24024
pkt_pts_time=0:00:00.800800
pkt_dts=24024
pkt_dts_time=0:00:00.800800
best_effort_timestamp=24024
best_effort_timestamp_time=0:00:00.800800
pkt_duration=1001
pkt_duration_time=0:00:00.033367
pkt_pos=33183
pkt_size=1.408203 Kibyte
width=640
height=360
pix_fmt=yuvj420p
sample_aspect_ratio=1:1
pict_type=P
coded_picture_number=21
display_picture_number=0
interlaced_frame=0
top_field_first=0
repeat_pict=0
color_range=pc
color_space=bt709
color_primaries=bt709
color_transfer=bt709
chroma_location=left
[/FRAME]
[FRAME]
media_type=video
stream_index=0
key_frame=0
pkt_pts=25025
pkt_pts_time=0:00:00.834167
pkt_dts=25025
pkt_dts_time=0:00:00.834167
best_effort_timestamp=25025
best_effort_timestamp_time=0:00:00.834167
pkt_duration=1001
pkt_duration_time=0:00:00.033367
pkt_pos=40297
pkt_size=885 byte
width=640
height=360
pix_fmt=yuvj420p
sample_aspect_ratio=1:1
pict_type=B
coded_picture_number=27
display_picture_number=0
interlaced_frame=0
top_field_first=0
repeat_pict=0
color_range=pc
color_space=bt709
color_primaries=bt709
color_transfer=bt709
chroma_location=left
[/FRAME]
[FRAME]
media_type=video
stream_index=0
key_frame=0
pkt_pts=26026
pkt_pts_time=0:00:00.867533
pkt_dts=26026
pkt_dts_time=0:00:00.867533
best_effort_timestamp=26026
best_effort_timestamp_time=0:00:00.867533
pkt_duration=1001
pkt_duration_time=0:00:00.033367
pkt_pos=38998
pkt_size=1.268555 Kibyte
width=640
height=360
pix_fmt=yuvj420p
sample_aspect_ratio=1:1
pict_type=B
coded_picture_number=26
display_picture_number=0
interlaced_frame=0
top_field_first=0
repeat_pict=0
color_range=pc
color_space=bt709
color_primaries=bt709
color_transfer=bt709
chroma_location=left
[/FRAME]
[FRAME]
media_type=video
stream_index=0
key_frame=0
pkt_pts=27027
pkt_pts_time=0:00:00.900900
pkt_dts=27027
pkt_dts_time=0:00:00.900900
best_effort_timestamp=27027
best_effort_timestamp_time=0:00:00.900900
pkt_duration=1001
pkt_duration_time=0:00:00.033367
pkt_pos=41182
pkt_size=967 byte
width=640
height=360
pix_fmt=yuvj420p
sample_aspect_ratio=1:1
pict_type=B
coded_picture_number=28
display_picture_number=0
interlaced_frame=0
top_field_first=0
repeat_pict=0
color_range=pc
color_space=bt709
color_primaries=bt709
color_transfer=bt709
chroma_location=left
[/FRAME]
[FRAME]
media_type=video
stream_index=0
key_frame=0
pkt_pts=28028
pkt_pts_time=0:00:00.934267
pkt_dts=N/A
pkt_dts_time=N/A
best_effort_timestamp=28028
best_effort_timestamp_time=0:00:00.934267
pkt_duration=1001
pkt_duration_time=0:00:00.033367
pkt_pos=37493
pkt_size=1.469727 Kibyte
width=640
height=360
pix_fmt=yuvj420p
sample_aspect_ratio=1:1
pict_type=P
coded_picture_number=25
display_picture_number=0
interlaced_frame=0
top_field_first=0
repeat_pict=0
color_range=pc
color_space=bt709
color_primaries=bt709
color_transfer=bt709
chroma_location=left
[/FRAME]
[FRAME]
media_type=video
stream_index=0
key_frame=0
pkt_pts=29029
pkt_pts_time=0:00:00.967633
pkt_dts=N/A
pkt_dts_time=N/A
best_effort_timestamp=29029
best_effort_timestamp_time=0:00:00.967633
pkt_duration=1001
pkt_duration_time=0:00:00.033367
pkt_pos=42149
pkt_size=1.225586 Kibyte
width=640
height=360
pix_fmt=yuvj420p
sample_aspect_ratio=1:1
pict_type=P
coded_picture_number=29
display_picture_number=0
interlaced_frame=0
top_field_first=0
repeat_pict=0
color_range=pc
color_space=bt709
color_primaries=bt709
color_transfer=bt709
chroma_location=left
[/FRAME]
[STREAM]
index=0
codec_name=h264
codec_long_name=H.264 / AVC / MPEG-4 AVC / MPEG-4 part 10
profile=High
codec_type=video
codec_tag_string=avc1
codec_tag=0x31637661
width=640
height=360
coded_width=640
coded_height=368
closed_captions=0
has_b_frames=2
sample_aspect_ratio=1:1
display_aspect_ratio=16:9
pix_fmt=yuvj420p
level=30
color_range=pc
color_space=bt709
color_transfer=bt709
color_primaries=bt709
chroma_location=left
field_order=unknown
refs=4
is_avc=true
nal_length_size=4
id=N/A
r_frame_rate=30000/1001
avg_frame_rate=30000/1001
time_base=1/30000
start_pts=0
start_time=0:00:00.000000
duration_ts=30030
duration=0:00:01.001000
bit_rate=346.501000 Kbit/s
max_bit_rate=N/A
bits_per_raw_sample=8
nb_frames=30
nb_read_frames=30
nb_read_packets=N/A
DISPOSITION:default=1
DISPOSITION:dub=0
DISPOSITION:original=0
DISPOSITION:comment=0
DISPOSITION:lyrics=0
DISPOSITION:karaoke=0
DISPOSITION:forced=0
DISPOSITION:hearing_impaired=0
DISPOSITION:visual_impaired=0
DISPOSITION:clean_effects=0
DISPOSITION:attached_pic=0
DISPOSITION:timed_thumbnails=0
TAG:language=und
TAG:handler_name=VideoHandler
TAG:vendor_id=[0][0][0][0]
[/STREAM]
[FORMAT]
filename=issue-3331.mp4
nb_streams=1
nb_programs=0
format_name=mov,mp4,m4a,3gp,3g2,mj2
format_long_name=QuickTime / MOV
start_time=0:00:00.000000
duration=0:00:01.001000
size=43.567383 Kibyte
bit_rate=356.547000 Kbit/s
probe_score=100
TAG:major_brand=isom
TAG:minor_version=512
TAG:compatible_brands=isomiso2avc1mp41
TAG:encoder=Lavf58.76.100
[/FORMAT]
```
</details>

The last frame has a `pkt_pts_time` value of 0:00:00.967633 which matches up with mpv's `time-pos` value (0.9676333333333333). This confirms qtiuto's report of a one frame gap.

This issue can also be seen in the mpv 0.34.0 GUI. Console output shows mpv pauses short of the end of the file:
```text
low-batt@gag issue-3331$ /Applications/mpv.app/Contents/MacOS/mpv --keep-open=always --osd-fractions issue-3331.mp4 
 (+) Video --vid=1 (*) (h264 640x360 29.970fps)
VO: [libmpv] 640x360 yuv420p
(Paused) V: 00:00:00.968 / 00:00:01.001 (97%)
```
In the GUI the progress bar stops short of the end:
![issue-3331-mpv-screen-shot](https://user-images.githubusercontent.com/86170219/141697108-37ef5bf1-652e-4a2a-9949-1131f6dd0e59.jpg)

### Fixing

Given this mpv behavior, IINA will have to take into account the value of the mpv `eof-reached` property when updating the `videoPosition` property of `PlaybackInfo`. If true then IINA will need to set `videoPosition` equal to `videoDuration` instead of the value of the mpv property `time-pos`.

I will shortly post a pull request containing that change.

**Comment by saagarjha** [comment]
We can fix this locally, but this sounds like we could also tell mpv to fix this? Thoughts on adding an other: upstream label on this?

**Comment by low-batt** [comment]
We can ask the mpv project about it, but my guess is that they won't want to change the behavior as it seems to be a "feature" of the `--keep-open` option for which the documentation refers to "seeking to the last frame". From the [mpv manual](https://mpv.io/manual/stable/):
>
**--keep-open=<yes|no|always>**
>
    Do not terminate when playing or seeking beyond the end of the file, and there is not next file to be played (and --loop is not used). Instead, pause the player. When trying to seek beyond end of the file, the player will attempt to seek to the last frame.

**Comment by saagarjha** [comment]
It's certainly worth trying; if the response isn't positive we can fix this on our side.

**Comment by low-batt** [comment]
Ok I will submit an issue. I'll try and get to it tomorrow. Was busy today getting Save panel done.

**Comment by low-batt** [comment]
The fix has merged into develop.

**Comment by low-batt** [comment]
Closing. Fixed in IINA 1.3.0.
